File: 리트코드/0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    public String serialize(TreeNode root) {
        if (root == null) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        sb.append("#,").append(root.val);

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            if (node.left != null) {
                sb.append(",").append(node.left.val);
                q.add(node.left);
            } else {
                sb.append(",#");
            }
            if (node.right != null) {
                sb.append(",").append(node.right.val);
                q.add(node.right);
            } else {
                sb.append(",#");
            }

        }

        return sb.toString();
    }

    public TreeNode deserialize(String data) {
        if (data.equals("")) {
            return null;
        }

        String[] split = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(split[1]));
        Queue<TreeNode> q = new LinkedList<>();


        q.add(root);
        int index = 2;
        while (!q.isEmpty()) {
            TreeNode node = q.poll();

            if (!split[index].equals("#")) {
                node.left = new TreeNode(Integer.parseInt(split[index]));
                q.add(node.left);
            }
            index += 1;
            
            if (!split[index].equals("#")) {
                node.right = new TreeNode(Integer.parseInt(split[index]));
                q.add(node.right);
            }
            index += 1;

        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: 리트코드/0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Codec() {
	// Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        if (root == null) {
            return ""
        }
        var sb: StringBuilder = StringBuilder()
        sb.append("#,").append(root.`val`)

        val q:Queue<TreeNode> = LinkedList()
        q.add(root)
        while (!q.isEmpty()) {
            val node = q.poll()
            if(node.left != null){
                sb.append(",").append(node.left!!.`val`)
                q.add(node.left)
            }else{
                sb.append(",#")
            }
            if(node.right != null){
                sb.append(",").append(node.right!!.`val`)
                q.add(node.right)
            }else{
                sb.append(",#")
            }
        }

        return sb.toString()
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        if (data == "") {
            return null
        }

        val nodes = data.split(",")

        val root:TreeNode = TreeNode(nodes[1].toInt())

        val q:Queue<TreeNode> = LinkedList()
        q.add(root)
        var index = 2
        while (!q.isEmpty()) {
            val node = q.poll()

            if (nodes[index] != "#") {
                node.left = TreeNode(nodes[index].toInt())
                q.add(node.left)
            }
            index++

            if (nodes[index] != "#") {
                node.right = TreeNode(nodes[index].toInt())
                q.add(node.right)
            }
            index++
        }
        return root
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * var ser = Codec()
 * var deser = Codec()
 * var data = ser.serialize(longUrl)
 * var ans = deser.deserialize(data)
 */